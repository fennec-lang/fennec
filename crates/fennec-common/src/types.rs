// Copyright 2023 Gregory Petrosyan <pgregory@pgregory.net>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

pub type HashMap<K, V> = std::collections::HashMap<K, V, ahash::RandomState>;
pub type HashSet<K> = std::collections::HashSet<K, ahash::RandomState>;

pub use crate::syncstate::*;
