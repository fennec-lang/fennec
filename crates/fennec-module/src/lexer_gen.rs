#![allow(unused_imports)]
#![allow(clippy::all)]
#![allow(clippy::pedantic)]
#[derive(Clone, Copy, PartialEq, Eq, Debug)]
pub(crate) enum LogosToken {
    Newline,
    Whitespace,
    KeywordModule,
    KeywordFennec,
    String,
    ErrorStringWithBackslashes,
    ErrorStringUnterminated,
    Comment,
}
impl<'s> ::logos::Logos<'s> for LogosToken {
    type Error = ();
    type Extras = ();
    type Source = str;
    fn lex(lex: &mut ::logos::Lexer<'s, Self>) {
        use logos::internal::{CallbackResult, LexerInternal};
        type Lexer<'s> = ::logos::Lexer<'s, LogosToken>;
        fn _end<'s>(lex: &mut Lexer<'s>) {
            lex.end()
        }
        fn _error<'s>(lex: &mut Lexer<'s>) {
            lex.bump_unchecked(1);
            lex.error();
        }
        static COMPACT_TABLE_0: [u8; 256] = [
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 0, 3, 3, 0, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 2, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
            3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
        ];
        macro_rules ! _fast_loop { ($ lex : ident , $ test : ident , $ miss : expr) => { while let Some (arr) = $ lex . read :: < & [u8 ; 16] > () { if $ test (arr [0]) { if $ test (arr [1]) { if $ test (arr [2]) { if $ test (arr [3]) { if $ test (arr [4]) { if $ test (arr [5]) { if $ test (arr [6]) { if $ test (arr [7]) { if $ test (arr [8]) { if $ test (arr [9]) { if $ test (arr [10]) { if $ test (arr [11]) { if $ test (arr [12]) { if $ test (arr [13]) { if $ test (arr [14]) { if $ test (arr [15]) { $ lex . bump_unchecked (16) ; continue ; } $ lex . bump_unchecked (15) ; return $ miss ; } $ lex . bump_unchecked (14) ; return $ miss ; } $ lex . bump_unchecked (13) ; return $ miss ; } $ lex . bump_unchecked (12) ; return $ miss ; } $ lex . bump_unchecked (11) ; return $ miss ; } $ lex . bump_unchecked (10) ; return $ miss ; } $ lex . bump_unchecked (9) ; return $ miss ; } $ lex . bump_unchecked (8) ; return $ miss ; } $ lex . bump_unchecked (7) ; return $ miss ; } $ lex . bump_unchecked (6) ; return $ miss ; } $ lex . bump_unchecked (5) ; return $ miss ; } $ lex . bump_unchecked (4) ; return $ miss ; } $ lex . bump_unchecked (3) ; return $ miss ; } $ lex . bump_unchecked (2) ; return $ miss ; } $ lex . bump_unchecked (1) ; return $ miss ; } return $ miss ; } while $ lex . test ($ test) { $ lex . bump_unchecked (1) ; } $ miss } ; }
        #[inline]
        fn goto6_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::KeywordModule));
        }
        #[inline]
        fn goto29_at1<'s>(lex: &mut Lexer<'s>) {
            match lex.read_at::<&[u8; 5usize]>(1usize) {
                Some(b"odule") => {
                    lex.bump_unchecked(6usize);
                    goto6_x(lex)
                }
                _ => _error(lex),
            }
        }
        #[inline]
        fn goto3_ctx3_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::Whitespace));
        }
        #[inline]
        fn pattern0(byte: u8) -> bool {
            match byte {
                9u8 | 32u8 => true,
                _ => false,
            }
        }
        #[inline]
        fn goto4_ctx3_x<'s>(lex: &mut Lexer<'s>) {
            _fast_loop!(lex, pattern0, goto3_ctx3_x(lex));
        }
        #[inline]
        fn goto2_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::Newline));
        }
        #[inline]
        fn goto28_at1<'s>(lex: &mut Lexer<'s>) {
            match lex.read_at::<&[u8; 1usize]>(1usize) {
                Some(&[10u8]) => {
                    lex.bump_unchecked(2usize);
                    goto2_x(lex)
                }
                _ => _error(lex),
            }
        }
        #[inline]
        fn goto7_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::KeywordFennec));
        }
        #[inline]
        fn goto30_at1<'s>(lex: &mut Lexer<'s>) {
            match lex.read_at::<&[u8; 5usize]>(1usize) {
                Some(b"ennec") => {
                    lex.bump_unchecked(6usize);
                    goto7_x(lex)
                }
                _ => _error(lex),
            }
        }
        #[inline]
        fn goto1_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::Newline));
        }
        #[inline]
        fn goto16_ctx16_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::ErrorStringUnterminated));
        }
        #[inline]
        fn pattern1(byte: u8) -> bool {
            COMPACT_TABLE_0[byte as usize] & 1 > 0
        }
        #[inline]
        fn goto12_ctx16_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::ErrorStringWithBackslashes));
        }
        #[inline]
        fn goto25_ctx16_x<'s>(lex: &mut Lexer<'s>) {
            let byte = match lex.read::<u8>() {
                Some(byte) => byte,
                None => return goto16_ctx16_x(lex),
            };
            match byte {
                byte if pattern1(byte) => {
                    lex.bump_unchecked(1usize);
                    goto25_ctx16_x(lex)
                }
                b'"' => {
                    lex.bump_unchecked(1usize);
                    goto12_ctx16_x(lex)
                }
                _ => goto16_ctx16_x(lex),
            }
        }
        #[inline]
        fn goto8_ctx16_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::String));
        }
        #[inline]
        fn goto24_ctx16_x<'s>(lex: &mut Lexer<'s>) {
            enum Jump {
                __,
                J25,
                J24,
                J8,
            }
            const LUT: [Jump; 256] = {
                use Jump::*;
                [
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, __, J24, J24, __, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J8, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J25, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                    J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24, J24,
                ]
            };
            let byte = match lex.read::<u8>() {
                Some(byte) => byte,
                None => return goto16_ctx16_x(lex),
            };
            match LUT[byte as usize] {
                Jump::J25 => {
                    lex.bump_unchecked(1usize);
                    goto25_ctx16_x(lex)
                }
                Jump::J24 => {
                    lex.bump_unchecked(1usize);
                    goto24_ctx16_x(lex)
                }
                Jump::J8 => {
                    lex.bump_unchecked(1usize);
                    goto8_ctx16_x(lex)
                }
                Jump::__ => goto16_ctx16_x(lex),
            }
        }
        #[inline]
        fn goto19_ctx19_x<'s>(lex: &mut Lexer<'s>) {
            lex.set(Ok(LogosToken::Comment));
        }
        #[inline]
        fn pattern2(byte: u8) -> bool {
            COMPACT_TABLE_0[byte as usize] & 2 > 0
        }
        #[inline]
        fn goto20_ctx19_x<'s>(lex: &mut Lexer<'s>) {
            _fast_loop!(lex, pattern2, goto19_ctx19_x(lex));
        }
        #[inline]
        fn goto27_at1<'s>(lex: &mut Lexer<'s>) {
            match lex.read_at::<&[u8; 1usize]>(1usize) {
                Some(b"/") => {
                    lex.bump_unchecked(2usize);
                    goto20_ctx19_x(lex)
                }
                _ => _error(lex),
            }
        }
        #[inline]
        fn goto31<'s>(lex: &mut Lexer<'s>) {
            enum Jump {
                __,
                J29,
                J4,
                J28,
                J30,
                J1,
                J24,
                J27,
            }
            const LUT: [Jump; 256] = {
                use Jump::*;
                [
                    __, __, __, __, __, __, __, __, __, J4, J1, __, __, J28, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, J4, __, J24, __, __, __,
                    __, __, __, __, __, __, __, __, __, J27, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, J30, __, __, __, __, __, __, J29, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                    __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __, __,
                ]
            };
            let byte = match lex.read::<u8>() {
                Some(byte) => byte,
                None => return _end(lex),
            };
            match LUT[byte as usize] {
                Jump::J29 => goto29_at1(lex),
                Jump::J4 => {
                    lex.bump_unchecked(1usize);
                    goto4_ctx3_x(lex)
                }
                Jump::J28 => goto28_at1(lex),
                Jump::J30 => goto30_at1(lex),
                Jump::J1 => {
                    lex.bump_unchecked(1usize);
                    goto1_x(lex)
                }
                Jump::J24 => {
                    lex.bump_unchecked(1usize);
                    goto24_ctx16_x(lex)
                }
                Jump::J27 => goto27_at1(lex),
                Jump::__ => _error(lex),
            }
        }
        goto31(lex)
    }
}
