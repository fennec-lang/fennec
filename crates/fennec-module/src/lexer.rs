// Copyright 2023 Gregory Petrosyan <pgregory@pgregory.net>
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use std::ops::Range;

use fennec_common::types::TextSize;
use logos::Logos as _;

use crate::lexer_gen::LogosToken;

#[derive(Clone, Copy, PartialEq, Eq)]
pub(crate) enum TokenKind {
    Newline,
    Whitespace,
    KwModule,
    KwFennec,
    String,
    Comment,
    Version,
    Error(TokenErrorKind),
    Eof,
}

impl TokenKind {
    pub fn is_trivia(self) -> bool {
        matches!(
            self,
            TokenKind::Whitespace | TokenKind::Comment | TokenKind::Error(_)
        )
    }
}

impl std::fmt::Debug for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenKind as T;
        let s = match *self {
            T::Newline => "LF",
            T::Whitespace => "WS",
            T::KwModule => "KW_MODULE",
            T::KwFennec => "KW_FENNEC",
            T::String => "STR",
            T::Comment => "COMMENT",
            T::Version => "VER",
            T::Error(err) => return write!(f, "ERR({err:?})"),
            T::Eof => "EOF",
        };
        write!(f, "{s}")
    }
}

impl std::fmt::Display for TokenKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenKind as T;
        let s = match *self {
            T::Newline => "newline",
            T::Whitespace => "space or tab",
            T::KwModule => "\"module\" keyword",
            T::KwFennec => "\"fennec\" keyword",
            T::String => "string",
            T::Comment => "comment",
            T::Version => "semantic version",
            T::Error(err) => return write!(f, "error ({err})"),
            T::Eof => "eof",
        };
        write!(f, "{s}")
    }
}

#[derive(Clone, Copy, PartialEq, Eq)]
pub(crate) enum TokenErrorKind {
    Identifier,
    SingleCarriageReturn,
    StringWithBackslashes,
    StringUnterminated,
    Other,
}

impl std::fmt::Debug for TokenErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenErrorKind as TE;
        let s = match self {
            TE::Identifier => "IDENT",
            TE::SingleCarriageReturn => "CR",
            TE::StringWithBackslashes => "STR_ESC",
            TE::StringUnterminated => "STR_TERM",
            TE::Other => "OTHER",
        };
        write!(f, "{s}")
    }
}

impl std::fmt::Display for TokenErrorKind {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        use TokenErrorKind as TE;
        let s = match self {
            TE::Identifier => "identifier",
            TE::SingleCarriageReturn => "carriage return (\\r)",
            TE::StringWithBackslashes => "string literal (with backslashes)",
            TE::StringUnterminated => "string literal (unterminated)",
            TE::Other => "other",
        };
        write!(f, "{s}")
    }
}

#[derive(Clone, PartialEq, Eq, Debug)]
pub(crate) struct Token {
    pub kind: TokenKind,
    pub len: TextSize,
}

impl Token {
    pub fn eof() -> Token {
        Token {
            kind: TokenKind::Eof,
            len: 0.into(),
        }
    }
}

const _: () = assert!(core::mem::size_of::<Token>() == 8); // just to be certain; it does not matter that much

fn to_token(token: Result<LogosToken, ()>, span: Range<usize>) -> Token {
    use LogosToken as LT;
    use TokenKind as T;
    let kind = match token {
        Ok(LT::Newline) => T::Newline,
        Ok(LT::ErrorSingleCarriageReturn) => T::Error(TokenErrorKind::SingleCarriageReturn),
        Ok(LT::Whitespace) => T::Whitespace,
        Ok(LT::KeywordModule) => T::KwModule,
        Ok(LT::KeywordFennec) => T::KwFennec,
        Ok(LT::String) => T::String,
        Ok(LT::ErrorStringWithBackslashes) => T::Error(TokenErrorKind::StringWithBackslashes),
        Ok(LT::ErrorStringUnterminated) => T::Error(TokenErrorKind::StringUnterminated),
        Ok(LT::Comment) => T::Comment,
        Ok(LT::ErrorIdentifier) => T::Error(TokenErrorKind::Identifier),
        Ok(LT::Version) => T::Version,
        Err(()) => T::Error(TokenErrorKind::Other),
    };
    let len: TextSize = (span.end - span.start)
        .try_into()
        .expect("token length must fit into 32 bits");
    Token { kind, len }
}

pub(crate) fn lex(input: &str) -> Vec<Token> {
    let mut tokens: Vec<Token> = Vec::new();
    let mut lexer = LogosToken::lexer(input);
    while let Some(tok) = lexer.next() {
        let token = to_token(tok, lexer.span());
        tokens.push(token);
    }
    tokens
}

#[cfg(test)]
mod tests {
    use fennec_common::types::{TextRange, TextSize};

    struct Tokens {
        input: String,
        tokens: Vec<super::Token>,
    }

    impl std::fmt::Debug for Tokens {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            let mut exploded = Vec::with_capacity(self.tokens.len());
            let mut start = TextSize::new(0);
            for tok in &self.tokens {
                let loc = TextRange::at(start, tok.len);
                exploded.push(((&self.input[loc]).to_owned(), tok.kind));
                start += tok.len;
            }
            exploded.fmt(f)
        }
    }

    fn lex(input: &str) -> Tokens {
        let tokens = super::lex(input);
        Tokens {
            input: input.to_owned(),
            tokens,
        }
    }

    #[test]
    fn lex_empty() {
        insta::assert_debug_snapshot!(lex(""));
    }

    #[test]
    fn lex_trivial_err() {
        insta::assert_debug_snapshot!(lex("?"));
    }

    #[test]
    fn lex_unterminated() {
        insta::assert_debug_snapshot!(lex("fennec \"hello\nmodule"));
    }

    #[test]
    fn lex_normal() {
        insta::assert_debug_snapshot!(lex(r#"
module "examples/hello"  // comment
fennec 0.1.0
"#));
    }
}
