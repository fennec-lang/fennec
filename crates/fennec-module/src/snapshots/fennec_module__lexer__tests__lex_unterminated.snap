---
source: crates/fennec-module/src/lexer.rs
expression: "lex(\"fennec \\\"hello\\nmodule\".to_owned())"
---
[
    (
        "fennec",
        KwFennec,
    ),
    (
        " ",
        Whitespace,
    ),
    (
        "\"hello",
        Error(
            StringUnterminated,
        ),
    ),
    (
        "\n",
        Newline,
    ),
    (
        "module",
        KwModule,
    ),
]
